File: py language/shape_service.py
# FILE: shape_system/shape_service.py
# LAYER: Service — the thin coordinator. No math. No I/O.

from datetime import datetime
from shapes import Shape          # From our Domain layer
from loggers import ShapeLogger   # From our Infrastructure layer


class ShapeService:
    """
    Receives a Shape, asks it to calculate its area,
    then hands the formatted result to the logger.
    That's it. Nothing else lives here.
    """

    def __init__(self, logger: ShapeLogger):
        if logger is None:
            raise ValueError("logger cannot be None.")
        self._logger = logger

    def process(self, shape: Shape) -> float:
        """Process one shape: calculate area, log it, return the area."""
        if shape is None:
            raise ValueError("shape cannot be None.")

        area      = shape.calculate_area()    # Domain does the math
        timestamp = datetime.now().strftime("%Y-%m-%d %H:%M:%S")
        entry     = f"[{timestamp}] {shape.shape_type}: area = {area:.4f}"
        self._logger.log(entry)               # Infrastructure does the output

        return area

File: js language/ShapeService.js
// FILE: shape-system/ShapeService.js
// LAYER: Service — thin coordinator. No math. No console. No document.

'use strict';

const { Shape }       = require('./shapes');    // Domain
const { ShapeLogger } = require('./loggers');   // Infrastructure abstraction

class ShapeService {
  /**
   * @param {ShapeLogger} logger  Any logger that extends ShapeLogger
   */
  constructor(logger) {
    if (!logger || typeof logger.log !== 'function') {
      throw new TypeError('logger must be a ShapeLogger instance.');
    }
    this._logger = logger;
  }

  /**
   * Calculates the area of the shape and logs the result.
   * @param {Shape} shape
   * @returns {number} the computed area
   */
  process(shape) {
    if (!(shape instanceof Shape)) {
      throw new TypeError('shape must be an instance of Shape.');
    }

    const area  = shape.calculateArea();           // Domain does the math
    const ts    = new Date().toISOString()
                    .replace('T', ' ').slice(0, 19);
    const entry = `[${ts}] ${shape.shapeType}: area = ${area.toFixed(4)}`;
    this._logger.log(entry);                       // Infrastructure does the output

    return area;
  }
}

module.exports = { ShapeService };